### DISCUSSION TITLE
How to get enum id to be written when it's on the outer struct?

### DISCUSSION THREAD

**Discussion by yanshay** [sharksforarms/deku#498] [Q&A]
In the below struct, the enum id is based on the outer struct field `data_type`.
When I write I want the `data_type` to be filled based on the enum variant in the `data` field.
How can I do that? Couldn't find an automated way and couldn't find how to access the deku id of the enum to fill it on update.

Also, while here, is there a way to place 'magic' inside a struct and not only as prefix to the struct? (that's the last field on the struct which is always 10). I could make a struct out of that field but seems like it should be simpler. 

```
#[derive(Debug, PartialEq, DekuRead, DekuWrite)]
#[deku(ctx ="data_type: u8", id = "data_type")]
enum ImprovWifiPacketData {
   #[deku(id=0x02)]
   RPC(u32)
}

#[derive(Debug, PartialEq, DekuRead, DekuWrite)]
#[deku(magic=b"IMPROV\x01")]
struct ImprovWifiPacket {
    data_type: u8,
    data_len: u8,
    #[deku(ctx = "*data_type")]
    data: ImprovWifiPacketData,
    checksum: u8,
    #[deku(update="10")]
    always_ten: u8,
}
```

**Comment by wcampbell0x2a** [comment]
The MR to add `ctx` to `update` is here: https://github.com/sharksforarms/deku/pull/415. I haven't work on it in awhile, I think you may have asked for something like this before in another issue!

The magic can just be an `assert` on a field in the struct.

**Reply by yanshay** [reply]
> The MR to add `ctx` to `update` is here: #415. I haven't work on it in awhile, I think you may have asked for something like this before in another issue!

I meant to do it w/o update.
What I did now was implement on the enum a `get_id` and I call it from the writer (and I do it for several enums which clutters the code). But this means the Id's are supplied twice, both to deku and in the `get_id` method.
If there was a way to have a `get_id` automatically generated by deku, or some attribute that fills in the value based on the id's I supply deku, it would simplify life and reduce a lot of clutter in my code.

> The magic can just be an `assert` on a field in the struct.

On write `assert` only verifies that something is there. Magic actually writes the data on its own.
So two solutions I could find: 
1. Use update to fill it (and then call update every time)
2. Create a struct with its own magic and put it there (which is what I chose so not to have to remember to call update every time)

Could be more elegant to have magic be a field attribute and not only a struct attribute but not that critical since there are reasonable alternatives already.

**Reply by wcampbell0x2a** [reply]
> Create a struct with its own magic and put it there (which is what I chose so not to have to remember to call update every time)

This is what I do, have a `const` value, then `assert` it for "magic".

**Reply by wcampbell0x2a** [reply]
> What I did now was implement on the enum a get_id and I call it from the writer (and I do it for several enums which clutters the code). But this means the Id's are supplied twice, both to deku and in the get_id method.

We do have an `deku_id` function: https://docs.rs/deku/latest/deku/trait.DekuEnumExt.html, It [wont be emitted for generics](https://github.com/sharksforarms/deku/blob/cbad15263e127a4e5b145e74d9543829c41633c8/deku-derive/src/macros/deku_read.rs#L473).

**Reply by yanshay** [reply]
Great, thanks. That's exactly what I was looking for.

**Reply by yanshay** [reply]
> > Create a struct with its own magic and put it there (which is what I chose so not to have to remember to call update every time)
> 
> This is what I do, have a `const` value, then `assert` it for "magic".

What do you mean by const? afaik rust doesn't support `const` struct fields. Maybe you refer to setting it to const in a `new` fn that initializes the struct? I may have several different `new` so I would need to set it in several places.

**Reply by wcampbell0x2a** [reply]
You're right, I push an MR that does "field magic": https://github.com/sharksforarms/deku/pull/503
